### PR TITLE
Foundation: Theme Provider

### PR DESCRIPTION
**Pull request by JasonGore**
#### Pull request checklist

- [x] Addresses an existing issue: #5979
- [x] Include a change request file using `$ npm run change`

#### Description of changes

Add Theme Provider to Foundation as specialized version of Customizer that simply takes in `scheme` as a string prop and takes care of applying appropriate theme data.

The approach taken here is that Theme Provider resides in Foundation and acts as an HOC to an injected versoin of Customizer. Customizer was modified to include a new optional transform prop which allows Theme Provider to supply the theme data according to the scheme prop provided.

This PR also removes the implicit `scheme` prop from createComponent. This means that all components that use Foundation will lose `scheme` as a prop which is a technically an "experimentally" breaking API change against those components.

As I'm on shield, I wanted to get this up for review on approach before buttoning up this PR 100% (of which all revolves around tests and minor cleanup.)

#### Focus areas to test

TODOs marked for unit tests updates

### CODE REVIEW

File: packages/utilities/src/Customizer.tsx [hunk 1]
```diff
@@ -96,6 +107,7 @@ export class Customizer extends BaseComponent<ICustomizerProps> {
   private _onCustomizationChange = () => this.forceUpdate();
 }
 
+// TODO: consider folding this back into Customizer if no longer used
 export function mergeCustomizations(props: ICustomizerProps, parentContext: ICustomizerContext): ICustomizerContext {
```
Comment (JasonGore) [inline packages/utilities/src/Customizer.tsx]: I exported this in #6391 and folding it back in would technically be an API breaking change, but I think it's good to discourage its use. I think it's too impactful on perf and I shouldn't have used it in my original approach.

File: packages/styling/src/styles/scheme.ts [hunk 2]
```diff
@@ -7,31 +7,26 @@ import { ISchemeNames, ITheme } from '../interfaces/index';
...
  * @param context - Context in which to get schemed customizations.
- * @param settingsTheme - Themes setting fallback if context does not have schemed theme.
  */
-export function getSchemedContext(
-  scheme: ISchemeNames,
-  context: ICustomizerContext,
-  settingsTheme?: ITheme
-): ICustomizerContext | undefined {
-  let newContext: ICustomizerContext | undefined;
+export function getSchemedContext(scheme: ISchemeNames, context: ICustomizerContext): ICustomizerContext {
```
Comment (JasonGore) [inline packages/styling/src/styles/scheme.ts]: I'm pretty sure this will get flagged as an API breaking change, but its type signature is backwards compatible (types have been tightened.)

File: packages/experiments/src/Foundation.ts [hunk 3]
```diff
@@ -75,13 +77,17 @@ export type IStatelessComponent<TComponentProps, TStyleSet extends IStyleSet<TSt
...
+const providers: IComponentProviders<any, any, any, ICustomizerContext, ITheme, ISchemeNames> = {
   mergeStyleSets,
-  getContextFromProps,
   getCustomizations,
   CustomizerContext
 };
 
+const themeProviders: IThemeProviders<ICustomizerContext, ITheme, ISchemeNames, ICustomizerProps> = {
+  getSchemedContext,
+  CustomizerComponent: Customizer
```
Comment (dzearing) [inline packages/experiments/src/Foundation.ts]: wondering if this is really needed to be configurable.

File: packages/utilities/src/Customizer.tsx [hunk 4]
```diff
@@ -56,7 +56,13 @@ export type ICustomizerProps = IBaseProps &
      * ```
      */
     scopedSettings: Settings | SettingsFunction;
-  }>;
+  }> & {
+    /**
+     * Optional transform function for context. Any implementations should take care to return context without
+     * mutating it.
+     */
+    contextPlugin?: (context: Readonly<ICustomizerContext>) => ICustomizerContext;
```
Comment (dzearing) [inline packages/utilities/src/Customizer.tsx]: contextTransform?

File: packages/foundation/src/themeProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,51 @@
+import * as React from 'react';
+
+export interface IThemeProviderProps<TScheme> {
+  scheme: TScheme;
```
Comment (dzearing) [inline packages/foundation/src/themeProvider.tsx]: would this also include theme?

File: packages/utilities/src/Customizer.tsx (hunk 4 above)
Comment (JasonGore) [reply packages/utilities/src/Customizer.tsx]: Yeah I think I agree with that name change.

File: packages/foundation/src/themeProvider.tsx (hunk 5 above)
Comment (JasonGore) [reply packages/foundation/src/themeProvider.tsx]: Well, I didn't add it here since `Customizer` handles that. We could add it here as a convenience but I didn't want to duplicate Customizer function off the bat.

**Comment by JasonGore** [comment]
We should talk about perf. I looked at how I could prevent `contextTransform` from being called whenever `Customizer` re-renders, but if `Customizer` is being forced to re-render, `contextTransform` is given the higher-level unmodified `context` every pass. This means `contextTransform` must modify the `context` every time it is called to apply the `scheme` or `theme` prop.

I think there are some bigger (or higher level) perf issues here:
1. `Customizer` is calling `mergeCustomizations` on every re-render, which is more expensive than context transform.
2. `styled` and `createComponent` are calling `Customizations.getSettings` on every render, which is also relatively more expensive.

Long story short, I wonder if this just adds fuel to the argument that we should remove global settings entirely? Maybe then we can look at how we can optimize merging settings without always merging in with global settings as `Customizations.getSettings` does now.

File: packages/styling/etc/styling.api.ts [hunk 6]
```diff
@@ -93,7 +93,7 @@ export function getIcon(name?: string): IIconRecord | undefined;
 export function getIconClassName(name: string): string;
 
 // @public
-export function getSchemedContext(scheme: ISchemeNames, context: ICustomizerContext, settingsTheme?: ITheme): ICustomizerContext | undefined;
+export function getSchemedContext(scheme: ISchemeNames, context: ICustomizerContext): ICustomizerContext;
```
Comment (JasonGore) [inline packages/styling/etc/styling.api.ts]: This is tightening the existing API function and is backwards compatible.
